Leptos 0.7 upgrade

ogghead opened the pull request:
Hi all!

This will ideally address #44. ~Creating this draft of Leptos 0.7 upgrade with partial changes -- several functions in lib.rs are still not fully updated and so as of time of creation, this does not compile. Creating this in hopes of getting eyes from Leptos experts on whether this is the right path to follow to upgrade. Contributions to move this to completion are appreciated from anyone interested!~

Update 11/4:
leptos_wasi now provides a universal foundation for running a Leptos back-end in Webassembly runtimes -- I have updated this PR with a draft template relying on that integration. Still need to update README with changes, but this is tentatively ready for use

Update 11/10:
Updated README and retested -- everything is working as expected -- ready for feedback!

ogghead committed:
Leptos 0.7 draft, not compiling

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

ogghead committed:
Merge branch 'main' into leptos-0.7

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

ogghead committed:
small updates to add new trait, update template

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

ogghead committed:
added hydration_context + lint fixes

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

ogghead committed:
small optimizations

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

ogghead committed:
Updated template using leptos_wasi integration

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

ogghead committed:
Rc1 working template

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

ogghead committed:
Readd arc

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

ogghead committed:
Use main leptos_wasi branch

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

ogghead committed:
readme updates

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

itowlson commented:
@benwis I can't add you as a reviewer for some reason but it would be great if you could also take a look - thanks!

itowlson commented on README.md:
```diff
@@ -2,7 +2,7 @@
 **THIS IS A WORK IN PROGRESS.** Actually 'in progress' probably oversells it. It is an early draft with a lot of learning as we go.
 
-This library provides integration services for running [Leptos](https://leptos-rs.github.io/leptos/) server-side applications in Spin.  It plays a role similar to Leptos' Actix and Axum integrations, bridging Spin's implementation of concepts and Leptos', and abstracting away common functionality to routing requests to Leptos views and server functions.
+This library provides integration services for running [Leptos](https://leptos-rs.github.io/leptos/) server-side applications in Spin. Previously, in Leptos versions 0.6 and below, a server integration similar to Leptos' Actix and Axum integrations was designed, published, and maintained in this repository (through the excellent work of @itowlson at Fermyon). As of Leptos 0.7, this integration re ...
```
If you're going to credit me, you should credit benwis in bigger letters - I started the project but he really made it sing.

ogghead replied:
Good callout, oversight on my part! I will update this now!

itowlson commented on templates/leptos-ssr/content/Cargo.toml:
```diff
@@ -9,29 +9,28 @@ edition = "2021"
+leptos_router = { version = "0.7.0-rc1" }
+leptos_wasi = { git = "https://github.com/leptos-rs/leptos_wasi", optional = true }
 spin-sdk = { version = "3", optional = true }
+wasi = { version = "=0.13.1", optional = true }
```
Will we be able to relax this bound a bit once Leptos is 0.7 RTM and `leptos_wasi` is on `crates.io`?

ogghead replied:
It has to do with the supported versions of wasi:http interface in Spin -- the latest minor version of wasi uses a version that (when I published this) was unsupported, though it's possible this is outdated now with the release of Spin 3.0

itowlson replied:
I checked in with one of the `wasi` crate maintainers and he believes that the crate (and indeed WASI interface) should be backward compatible with WASI 0.2.0.  If it's not, let me know and I will rope him in.  Thanks!

ogghead replied:
Interesting! Let me give it another shot without this exact dep, I will update shortly

ogghead replied:
~Indeed this appears to run just fine for me now -- not sure what exactly was going wrong before! Updated to use 0.13 without specific version~

When I run with `version = "0.13.3+wasi-0.2.2"`, I see
```
2024-11-19T02:11:57.756986Z ERROR spin_trigger::cli: Trigger executor failed
Error: Expected component to export one of `wasi:http/incoming-handler@0.2.0-rc-2023-10-18`, `wasi:http/incoming-handler@0.2.0-rc-2023-11-10`, `wasi:http/incoming-handler@0.2.0`, `wasi:http/incoming-handler@0.2.1`, or `fermyon:spin/inbound-http` but it exported none of those
```

The handler that _is_ exported is `wasi:http/incoming-handler@0.2.2` in the latest minor version of `wasi`, but perhaps I am missing something

itowlson replied:
Oh!  I wonder if this relates to Spin checking whether it should apply an adapter.  It knows the RCs need adapting, but perhaps it doesn't yet know that 0.2.2 _doesn't_ need adapting.  Don't worry then, leave as is until we can fix in Spin!  Sorry for the bad guidance!

ogghead replied:
All good! I should have documented this in a comment to help with context -- if you by chance know of a tracking ticket in Spin I can link that on the dependency (and totally no worries either way, I can link the error for now)

itowlson replied:
Tracking at https://github.com/fermyon/spin/issues/2928

ogghead replied:
Updated with the link, thank you for making that! I would be interested in helping to address that, though I must admit my knowledge of that facet of Spin is quite limited

itowlson replied:
I'm sure one of our contributors who's more knowledgeable than I about that area would be delighted to point you in the right direction and support you learning the ways of it.  Feel free to put your hand up if you have the time and enthusiasm!

itowlson commented on templates/leptos-ssr/content/src/app.rs:
```diff
@@ -1,35 +1,57 @@
 
-      // sets the document title
-      <Title text="Welcome to Leptos"/>
+        <Title text="Welcome to Leptos X WebAssembly!"/>
```
If we wish to change this I would suggest Spin rather than WebAssembly (because Leptos uses WebAssembly on the client anyway).

ogghead replied:
Ahh, this was some copy pasta from merging the templates. Will update now!

itowlson commented on templates/leptos-ssr/content/src/server.rs:
```diff
@@ -1,19 +1,43 @@
 
-    let mut routes = RouteTable::build(app);
-    routes.add_server_fn_prefix("/api").unwrap();
+impl Guest for LeptosServer {
```
I wonder if we should use the `#[http_component]` macro (which allows the IncomingRequest/ResponseOutparam signature) instead of the raw binding here?  It may be more familiar to Spin users.  But maybe app developers don't touch this file enough to make any odds.

ogghead replied:
I ran into some challenges with reconciling the types before when I dug into making that work -- I hit a dead-end trying to convert a Spin IncomingRequest into the one expected by leptos_wasi iirc. I will take another crack at this

itowlson replied:
@ogghead Oh, if it doesn't work automatically then don't worry about it.  It's no biggie.

ogghead replied:
Yeah, still running into challenges around converting types sadly -- I had gone down the route of newtypes to satisfy orphan rule before but it was becoming a rabbithole. This would be ideal though, I can create an issue to track that effort if you're on board with that

itowlson replied:
Yeah it is definitely not worth investing more time in.  Again, sorry for sending you in an unsuccessful direction.

ogghead replied:
Totally no worries! I spent more time than I would care to admit trying to get that working a couple weeks ago XD I do appreciate the feedback, as I did want to test whether ~Rohan would come to our aid~ Spin 3 or a new version of leptos_wasi might get this rolling

itowlson commented on templates/leptos-ssr/content/src/server.rs:
```diff
@@ -1,19 +1,43 @@
+
+    Handler::build(request, response_out)
+        .expect("could not create handler")
+        .with_server_fn::<SaveCount>()
```
It's worth commenting that developers need to add new server functions here.  It's quite easy to miss now it's in the middle of a fluent chain.

ogghead replied:
Good point, I will add this now

itowlson commented on templates/leptos-ssr/content/src/server.rs:
```diff
@@ -1,19 +1,43 @@
+    let leptos_options = conf.leptos_options;
+
+    Handler::build(request, response_out)
+        .expect("could not create handler")
```
I'm a bit wary of using `expect` rather than propagating an error, but maybe there's nothing we can meaningfully do with the error?  Still, I'd prefer to centralise the panic assumption further up - could we have this function return a `Result<()>` and process the falure case in `handle`?

ogghead replied:
Very good point -- handling behavior with Result is more resilient and Rustic. I will update this now

itowlson commented:
Wow, leptos-wasi makes all this _soooooooooo_ much simpler...!

itowlson commented:
@ogghead Just had a query on Discord about this.  Would you like this merged at stands with Leptos 7-rc1 (once Ben has looked and any required tweaks/tidying have been done), or were you envisaging waiting for Leptos 7 release?  I am okay with either, just checking on how you feel.

ogghead commented:
> @ogghead Just had a query on Discord about this.  Would you like this merged at stands with Leptos 7-rc1 (once Ben has looked and any required tweaks/tidying have been done), or were you envisaging waiting for Leptos 7 release?  I am okay with either, just checking on how you feel.
> 
> 

Hey great timing 👋 thanks for reviewing! I think this is _tentatively_ ready to merge barring any additional required tweaks/tidying. I am not sure exactly when the planned release date of Leptos 0.7 is but I would hope any additional changes needed would be minimal. So, merging with these dependencies and creating a quick follow-up PR to update versions + release a tagged version of the template once 0.7 is out makes sense to me!

ogghead committed:
Minor changes to address PR feedback, lock leptos-wasi to commit

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

ogghead commented:
Ok, addressed PR feedback thus far, but please let me know if you have any other thoughts @itowlson! I very much appreciate your feedback on this :) @benwis hopefully this PR is up to snuff!

itowlson commented on templates/leptos-ssr/content/src/server.rs:
```diff
@@ -1,19 +1,53 @@
 
-    render_best_match_to_stream(req, resp_out, &routes, app, &conf.leptos_options).await
+    let mut conf = get_configuration(None).unwrap();
+    conf.leptos_options.output_name = Arc::from("newleptos".to_owned());
```
Should `newleptos` be a template-derived app name here?  cf previous `conf.leptos_options.output_name = "{{project-name | snake_case}}".to_owned();`

ogghead replied:
Ahh good catch, I copied some non-templated parts back when I implemented changes, apologies for not reviewing more closely. I will fix now

itowlson reviewed (approved):
LGTM except for one detail in the template - thanks!

ogghead committed:
Revert templating removal, add comment linking to Spin tracking issue on patch version

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

itowlson commented:
@benwis Sorry to badger, but do you know if you'll have bandwidth to look at this?  No worries if not.

benwis commented on templates/leptos-ssr/content/src/app.rs:
```diff
@@ -1,35 +1,57 @@
+
     view! {
-      <Stylesheet id="leptos" href="/pkg/{{project-name | snake_case}}.css"/>
+        <Meta name="charset" content="UTF-8"/>
```
Is there no css file in this one? It's fine if not, but if so it's removed it. It might be better to keep it

ogghead replied:
I moved this to the [shell function](https://github.com/fermyon/leptos-spin/pull/48/files#diff-782be6a82ee08bd1b4914a1c0ca4dba43937f23e80b5d065f06287678ef51ee1R15) -- I recall that was the pattern I saw in some Leptos examples but I am certainly not a Leptos expert! Let me know if it is preferable to keep the stylesheet in App

benwis replied:
That's a good way to do it, I missed that

benwis commented on README.md:
```diff
@@ -49,45 +49,28 @@ Now the app should be served at `http://127.0.0.1:3000`
 
 * All server functions (`#[server]`) **must** be explicitly registered (see usage sample below). In native code, Leptos uses a clever macro to register them automatically; unfortunately, that doesn't work in WASI.
 
-* When using a context value in a component in a `feature = "ssr"` block, you **must** call `use_context` **not** `expect_context`. `expect_context` will panic during routing.  E.g.
```
Did you remove this section for a reason?

ogghead replied:
I was seeing this work with expect_context in testing -- for example [in the NotFound component](https://github.com/fermyon/leptos-spin/pull/48/files#diff-782be6a82ee08bd1b4914a1c0ca4dba43937f23e80b5d065f06287678ef51ee1R81)

Perhaps this is still unsupported though and I was misreading that behavior?

benwis replied:
We definitely still do route generation that could not have it, but depending on what you're getting it may or may not work. Request or something from Request is the most likely to hit this issue. Assuming it still is, this could be reworded and clarified

ogghead replied:
Ah good to know! Updated to re-add this section with a bit of additional context

benwis commented on templates/leptos-ssr/content/src/server.rs:
```diff
@@ -1,19 +1,53 @@
 
-    render_best_match_to_stream(req, resp_out, &routes, app, &conf.leptos_options).await
+    let mut conf = get_configuration(None).unwrap();
+    conf.leptos_options.output_name = Arc::from("{{project-name | snake_case}}".to_owned());
```
This is kinda strange, cargo-leptos should be setting this as an env var, and get_configuration(None) should load it you shouldn't be needing to hardcode it

itowlson replied:
@benwis This might be historical (it's existing code which GitHub is getting confused and showing as a diff).  During early integration work I found myself needing to force the project name, and I tried a bunch of ways to do that: it's quite likely more of them got left in than are strictly needed...!

ogghead replied:
Good callout, a test Leptos Spin app does seem to function fine without this -- I went ahead and removed here and in README.

Thanks for the context @itowlson -- do you remember why the project name had to be forced?

itowlson replied:
@ogghead I think it was to do with finding assets under the `/pkg` route - it was looking for assets using the wrong name.  But yeah this was very much an "at the time" thing.  I am happy to write it off as an accident of history that is no longer needed.  Thanks for testing and thanks @benwis for querying it!

benwis commented on README.md:
```diff
@@ -49,45 +49,28 @@ Now the app should be served at `http://127.0.0.1:3000`
+    use leptos_wasi::prelude::Handler;
+
+    let mut conf = get_configuration(None).unwrap();
+    conf.leptos_options.output_name = Arc::from("testleptos".to_owned());
```
This whole configuration thing, much like the other one, is somewhat bypassing cargo-leptos's values for dev and the env vars people usually set for prod

ogghead replied:
That's fair -- I think in the context of WASI we may need to configure env vars in a different way, but this could just be a remnant -- I will investigate cleaning these up to match the typical Leptos patterns

ogghead replied:
Indeed a test app seemed to work fine without this override. Pushed up the removal here + in codegen, good callout!

benwis reviewed (commented):
Overall pretty good, left a couple comments

ogghead committed:
Remove hardcoded output_name

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

ogghead commented:
Thanks for the review @benwis and the bump @itowlson! I have responded to comments + pushed up changes to address. I will keep an eye out for additional feedback + discussion on open comments -- I appreciate the eyes on this PR!

ogghead committed:
Update README for expect_context

Signed-off-by: Darwin Boersma <darwin@sadlark.com>

ogghead commented:
I appreciate the quick review and additional context @benwis! Updated to address that comment now

benwis reviewed (approved):
LGTM

itowlson commented:
Thanks @ogghead for your work on this and @benwis for bringing your knowledge and attention to detail!  Hopefully future updates will be easier now that Leptos has a WASI integration within the org!
